README.md
<!--suppress HtmlDeprecatedAttribute, HttpUrlsUsage -->

<div align="center">
  <p>
    <img src="https://s1.imagehub.cc/images/2023/03/07/a611060ac75cf0d48edacff319cc1666.png" alt="autojs6-documentation-banner_800×208_transparent" border="0" width="660"/>
  </p>

  <p>AutoJs6 应用文档</p>

  <p>
    <a href="http://docs-project.autojs6.com/blob/master/project.json"><img alt="Version name" src="https://img.shields.io/badge/dynamic/json?color=1283C3&label=version&query=%24.versionName&url=https%3A%2F%2Fraw.githubusercontent.com%2FSuperMonster003%2FAutoJs6-Documentation%2Fmaster%2Fproject.json"/></a>
    <a href="http://docs-project.autojs6.com/issues"><img alt="GitHub closed issues" src="https://img.shields.io/github/issues/SuperMonster003/AutoJs6-Documentation?color=009688"/></a>
    <a href="http://docs-project.autojs6.com/commit/f7389a70ffc7bac4406cd0b19a1e6341d6e50238"><img alt="Created" src="https://img.shields.io/date/1657941168?color=2e7d32&label=created"/></a>
    <br>
    <a href="http://docs-project.autojs6.com/find/master"><img alt="GitHub Code Size" src="https://img.shields.io/github/languages/code-size/SuperMonster003/AutoJs6-Documentation?color=795548"/></a>
    <a href="http://docs-project.autojs6.com/blob/master/LICENSE"><img alt="GitHub License" src="https://img.shields.io/github/license/SuperMonster003/AutoJs6-Documentation?color=534BAE"/></a>
    <a href="https://www.jetbrains.com/?from=Ant-Forest"><img alt="JetBrains supporter" src="https://img.shields.io/badge/supporter-JetBrains-ee4677"/></a>
  </p>
</div>

******

### 简介

******

- [AutoJs6](http://project.autojs6.com) 应用文档
- 克隆 (Clone) 自 [hyb1996/AutoJs-Docs](https://github.com/hyb1996/AutoJs-Docs/)
- 模板 / 样式 / Generator 来自 [Node.js](https://github.com/nodejs/node/tree/master/doc/)

******

### 阅读

******

* [点此阅读 (网页版)](https://docs.autojs6.com)

******

### 说明

******

- 文档可能会随时更新
- 部分文档与实际代码行为可能存在出入
- 如有任何问题可在当前项目的 [议题 (Issues)](http://docs-project.autojs6.com/issues) 页面提交反馈
- 如需快速了解文档的常见问题, 可参阅 [疑难解答](https://docs.autojs6.com/#/qa) 章节

******

### 进度

******

- 基于 `Auto.js 4.1.0` 文档修改完善, [部署进度](https://docs.autojs6.com/#/progress) 可能相对缓慢

******

### 版本历史

******

[comment]: <> "Version history only shows last 3 versions"

## v1.1.8

<p style="font: bold 0.8em sans-serif; color: #888888">2023/12/01</p>

- `新增` [中文转换 (OpenCC)](https://docs.autojs6.com/#/opencc) 文档
- `新增` [OpenCCConversion](https://docs.autojs6.com/#/openCCConversionType) 类型
- `新增` [选择器](https://docs.autojs6.com/#/uiSelectorType) 章节增加 [plus](https://docs.autojs6.com/#/uiObjectType?id=m-plus) / [append](https://docs.autojs6.com/#/uiObjectType?id=m-append) 条目
- `新增` [控制台 (Console)](https://docs.autojs6.com/#/console) 章节增加 [setTouchable](https://docs.autojs6.com/#/console?id=m-settouchable) 条目
- `新增` [ConsoleBuildOptions](https://docs.autojs6.com/#/consoleBuildOptionsType) 章节增加 [touchable](https://docs.autojs6.com/#/consoleBuildOptionsType?id=p-touchable) 条目
- `优化` [光学字符识别 (OCR)](https://docs.autojs6.com/#/ocr) 章节增加 Paddle 工作模式使用提示
- `优化` 完善 [Shizuku](https://docs.autojs6.com/#/shizuku) 章节
- `优化` 完善 [选择器](https://docs.autojs6.com/#/uiSelectorType) 章节

## v1.1.7

<p style="font: bold 0.8em sans-serif; color: #888888">2023/10/30</p>

- `新增` [Shizuku](https://docs.autojs6.com/#/shizuku) 文档
- `新增` [WebSocket](https://docs.autojs6.com/#/websocketType) 文档
- `新增` [条码 (Barcode)](https://docs.autojs6.com/#/barcode) 文档
- `新增` [二维码 (QR Code)](https://docs.autojs6.com/#/qrcode) 文档
- `优化` 完善 [颜色 (Color)](https://docs.autojs6.com/#/color) 章节
- `优化` 完善 [光学字符识别 (OCR)](https://docs.autojs6.com/#/ocr) 章节

## v1.1.6

<p style="font: bold 0.8em sans-serif; color: #888888">2023/07/21</p>

- `优化` 完善 [控件节点](https://docs.autojs6.com/#/uiObjectType) 章节

##### 更多版本历史可参阅

* [CHANGELOG.md](http://docs-project.autojs6.com/blob/master/api/changelog.md)

changelog.md
# 文档更新日志 (Changelog)

## v1.1.8

<p style="font: bold 0.8em sans-serif; color: #888888">2023/12/01</p>

- `新增` [中文转换 (OpenCC)](https://docs.autojs6.com/#/opencc) 文档
- `新增` [OpenCCConversion](https://docs.autojs6.com/#/openCCConversionType) 类型
- `新增` [选择器](https://docs.autojs6.com/#/uiSelectorType) 章节增加 [plus](https://docs.autojs6.com/#/uiObjectType?id=m-plus) / [append](https://docs.autojs6.com/#/uiObjectType?id=m-append) 条目
- `新增` [控制台 (Console)](https://docs.autojs6.com/#/console) 章节增加 [setTouchable](https://docs.autojs6.com/#/console?id=m-settouchable) 条目
- `新增` [ConsoleBuildOptions](https://docs.autojs6.com/#/consoleBuildOptionsType) 章节增加 [touchable](https://docs.autojs6.com/#/consoleBuildOptionsType?id=p-touchable) 条目
- `优化` [光学字符识别 (OCR)](https://docs.autojs6.com/#/ocr) 章节增加 Paddle 工作模式使用提示
- `优化` 完善 [Shizuku](https://docs.autojs6.com/#/shizuku) 章节
- `优化` 完善 [选择器](https://docs.autojs6.com/#/uiSelectorType) 章节

## v1.1.7

<p style="font: bold 0.8em sans-serif; color: #888888">2023/10/30</p>

- `新增` [Shizuku](https://docs.autojs6.com/#/shizuku) 文档
- `新增` [WebSocket](https://docs.autojs6.com/#/websocketType) 文档
- `新增` [条码 (Barcode)](https://docs.autojs6.com/#/barcode) 文档
- `新增` [二维码 (QR Code)](https://docs.autojs6.com/#/qrcode) 文档
- `优化` 完善 [颜色 (Color)](https://docs.autojs6.com/#/color) 章节
- `优化` 完善 [光学字符识别 (OCR)](https://docs.autojs6.com/#/ocr) 章节

## v1.1.6

<p style="font: bold 0.8em sans-serif; color: #888888">2023/07/21</p>

- `优化` 完善 [控件节点](https://docs.autojs6.com/#/uiObjectType) 章节

## v1.1.5

<p style="font: bold 0.8em sans-serif; color: #888888">2023/07/06</p>

- `新增` [密文 (Crypto)](https://docs.autojs6.com/#/crypto) 文档
- `新增` [CryptoCipherOptions](https://docs.autojs6.com/#/cryptoCipherOptionsType) / [CryptoKey](https://docs.autojs6.com/#/cryptoKeyType) / [CryptoKeyPair](https://docs.autojs6.com/#/cryptoKeyPairType) 等类型
- `修复` floaty 模块 widht 拼写失误 _[`issue #1`](http://docs-project.autojs6.com/issues/1)_
- `优化` 完善 [Base64](https://docs.autojs6.com/#/base64) 章节
- `优化` 完善 [颜色 (Color)](https://docs.autojs6.com/#/color) 章节

## v1.1.4

<p style="font: bold 0.8em sans-serif; color: #888888">2023/05/26</p>

- `新增` [console.resetGlobalLogConfig](https://docs.autojs6.com/#/console?id=m-resetgloballogconfig) 文档
- `新增` [web.newWebSocket](https://docs.autojs6.com/#/web?id=m-newwebsocket) 文档
- `优化` 完善 [全能类型 (Omnipotent Types)](https://docs.autojs6.com/#/omniTypes) 章节
- `优化` 完善 [安卓 API 级别 (Android API Level)](https://docs.autojs6.com/#/apiLevel) 章节

## v1.1.3

<p style="font: bold 0.8em sans-serif; color: #888888">2023/04/29</p>

- `新增` [颜色类 (Color)](https://docs.autojs6.com/#/colorType) 文档
- `新增` [控制台 (Console)](https://docs.autojs6.com/#/console) 文档
- `新增` [标准化 (Standardization)](https://docs.autojs6.com/#/s13n) 文档
- `新增` [全能类型 (Omnipotent Types)](https://docs.autojs6.com/#/omniTypes) 文档
- `新增` [NoticeBuilder](https://docs.autojs6.com/#/noticeBuilderType) / [NoticeChannelOptions](https://docs.autojs6.com/#/noticeChannelOptionsType) / [NoticeOptions](https://docs.autojs6.com/#/noticeOptionsType) 等类型
- `新增` 示例代码区域增加 Copy 按钮以复制代码内容
- `新增` 文档中的图片内容支持点击以全屏方式查看
- `修复` 文档内容中部分图片资源丢失的问题
- `优化` 生成器根据 properties 文件自动获取 AutoJs6 版本信息
- `优化` 压缩本地 JavaScript 文件以提升页面加载速度
- `优化` 本地化字体文件避免网络条件不佳时影响页面加载速度
- `优化` 部分表格内容强制禁用自动断行以提升阅读体验
- `优化` 完善 [颜色 (Color)](https://docs.autojs6.com/#/color) 章节
- `优化` 完善 [消息通知 (Notice)](https://docs.autojs6.com/#/notice) 章节
- `优化` 完善 [光学字符识别 (OCR)](https://docs.autojs6.com/#/ocr) 章节

## v1.1.2

<p style="font: bold 0.8em sans-serif; color: #888888">2023/03/21</p>

- `新增` [光学字符识别 (OCR)](https://docs.autojs6.com/#/ocr) 文档
- `新增` [消息通知 (Notice)](https://docs.autojs6.com/#/notice) 文档
- `新增` [HttpRequestHeaders](https://docs.autojs6.com/#/httpRequestHeadersType) / [HttpResponseHeaders](https://docs.autojs6.com/#/httpResponseHeadersType) / [OpenCVRect](https://docs.autojs6.com/#/opencvRectType) 等类型
- `新增` [通知渠道](https://docs.autojs6.com/#/glossaries?id=通知渠道) / [HTTP 标头](https://docs.autojs6.com/#/glossaries?id=HTTP-标头) / [MIME 类型](https://docs.autojs6.com/#/glossaries?id=MIME-类型) / [HTTP 请求方法](https://docs.autojs6.com/#/glossaries?id=HTTP-请求方法) 等术语
- `新增` [颜色 (Color)](https://docs.autojs6.com/#/color) 章节增加 [toColorStateList](https://docs.autojs6.com/#/color?id=m-tocolorstatelist) 及 [setPaintColor](https://docs.autojs6.com/#/color?id=m-setpaintcolor) 条目
- `修复` 文档更新日志条目中的链接无效的问题
- `优化` 完善 [疑难解答 (Q & A)](https://docs.autojs6.com/#/qa) 章节

## v1.1.1

<p style="font: bold 0.8em sans-serif; color: #888888">2023/03/02</p>

- `新增` [Base64](https://docs.autojs6.com/#/base64) 文档
- `新增` [活动 (Activity)](https://docs.autojs6.com/#/activity) 文档
- `新增` [插件 (Plugins)](https://docs.autojs6.com/#/plugins) 文档
- `新增` [存储 (Storages)](https://docs.autojs6.com/#/storages) 文档
- `新增` [万维网 (Web)](https://docs.autojs6.com/#/web) 文档
- `新增` [global.species](https://docs.autojs6.com/#/global?id=m-species) 文档
- `新增` [术语](https://docs.autojs6.com/#/glossaries) 章节增加 [阈值](https://docs.autojs6.com/#/glossaries?id=阈值) / [注入](https://docs.autojs6.com/#/glossaries?id=注入) 等条目
- `新增` [数据类型](https://docs.autojs6.com/#/dataTypes) 章节增加 [Storage](https://docs.autojs6.com/#/storageType) / [ColorDetectionAlgorithm](https://docs.autojs6.com/#/dataTypes?id=colordetectionalgorithm) / [InjectableWebView](https://docs.autojs6.com/#/injectableWebViewType) 等类型
- `修复` 示例代码中与美元符号 ($) 相关内容可能出现占位符替换失败的问题

static/search.js
(function () {
    /**
     * Converts a colon formatted string to a object with properties.
     *
     * This is process a provided string and look for any tokens in the format
     * of `:name[=value]` and then convert it to a object and return.
     * An example of this is ':include :type=code :fragment=demo' is taken and
     * then converted to:
     *
     * ```
     * {
     *  include: '',
     *  type: 'code',
     *  fragment: 'demo'
     * }
     * ```
     *
     * @param {string}   str   The string to parse.
     *
     * @return {object}  The original string and parsed object, { str, config }.
     */
    function getAndRemoveConfig(str) {
        if (str === void 0) str = '';

        var config = {};

        if (str) {
            str = str
                .replace(/^('|")/, '')
                .replace(/('|")$/, '')
                .replace(/(?:^|\s):([\w-]+:?)=?([\w-%]+)?/g, function (m, key, value) {
                    if (key.indexOf(':') === -1) {
                        config[key] = (value && value.replace(/&quot;/g, '')) || true;
                        return '';
                    }

                    return m;
                })
                .trim();
        }

        return { str: str, config: config };
    }

    /* eslint-disable no-unused-vars */

    var INDEXS = {};

    var LOCAL_STORAGE = {
        EXPIRE_KEY: 'docsify.search.expires',
        INDEX_KEY: 'docsify.search.index',
    };

    function resolveExpireKey(namespace) {
        return namespace
            ? ((LOCAL_STORAGE.EXPIRE_KEY) + '/' + namespace)
            : LOCAL_STORAGE.EXPIRE_KEY;
    }

    function resolveIndexKey(namespace) {
        return namespace
            ? ((LOCAL_STORAGE.INDEX_KEY) + '/' + namespace)
            : LOCAL_STORAGE.INDEX_KEY;
    }

    function escapeHtml(string) {
        var entityMap = {
            '&': '&amp;',
            '<': '&lt;',
            '>': '&gt;',
            '"': '&quot;',
            '\'': '&#39;',
        };

        return String(string).replace(/[&<>"']/g, function (s) {
            return entityMap[s];
        });
    }

    function getAllPaths(router) {
        var paths = [];

        Docsify.dom
            .findAll('.sidebar-nav a:not(.section-link):not([data-nosearch])')
            .forEach(function (node) {
                var href = node.href;
                var originHref = node.getAttribute('href');
                var path = router.parse(href).path;

                if (
                    path &&
                    paths.indexOf(path) === -1 &&
                    !Docsify.util.isAbsolutePath(originHref)
                ) {
                    paths.push(path);
                }
            });

        return paths;
    }

    function getTableData(token) {
        if (!token.text && token.type === 'table') {
            token.cells.unshift(token.header);
            token.text = token.cells
                .map(function (rows) {
                    return rows.join(' | ');
                })
                .join(' |\n ');
        }
        return token.text;
    }

    function getListData(token) {
        if (!token.text && token.type === 'list') {
            token.text = token.raw;
        }
        return token.text;
    }

    function saveData(maxAge, expireKey, indexKey) {
        localStorage.setItem(expireKey, Date.now() + maxAge);
        localStorage.setItem(indexKey, JSON.stringify(INDEXS));
    }

    function genIndex(path, content, router, depth) {
        if (content === void 0) content = '';

        var tokens = window.marked.lexer(content);
        var slugify = window.Docsify.slugify;
        var index = {};
        var slug;
        var title = '';

        tokens.forEach(function (token, tokenIndex) {
            if (token.type === 'heading' && token.depth <= depth) {
                var ref = getAndRemoveConfig(token.text);
                var str = ref.str;
                var config = ref.config;

                if (config.id) {
                    slug = router.toURL(path, { id: slugify(config.id) });
                } else {
                    slug = router.toURL(path, { id: slugify(escapeHtml(token.text)) });
                }

                if (str) {
                    title = str
                        .replace(/<!-- {docsify-ignore} -->/, '')
                        .replace(/{docsify-ignore}/, '')
                        .replace(/<!-- {docsify-ignore-all} -->/, '')
                        .replace(/{docsify-ignore-all}/, '')
                        .trim();
                }

                index[slug] = { slug: slug, title: title, body: '' };
            } else {
                if (tokenIndex === 0) {
                    slug = router.toURL(path);
                    index[slug] = {
                        slug: slug,
                        title: path !== '/' ? path.slice(1) : 'Home Page',
                        body: token.text || '',
                    };
                }

                if (!slug) {
                    return;
                }

                if (!index[slug]) {
                    index[slug] = { slug: slug, title: '', body: '' };
                } else if (index[slug].body) {
                    token.text = getTableData(token);
                    token.text = getListData(token);

                    index[slug].body += '\n' + (token.text || '');
                } else {
                    token.text = getTableData(token);
                    token.text = getListData(token);

                    index[slug].body = index[slug].body
                        ? index[slug].body + token.text
                        : token.text;
                }
            }
        });
        slugify.clear();
        return index;
    }

    function ignoreDiacriticalMarks(keyword) {
        if (keyword && keyword.normalize) {
            return keyword.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
        }
        return keyword;
    }

    /**
     * @param {String} query Search query
     * @returns {Array} Array of results
     */
    function search(query) {
        var matchingResults = [];
        var data = [];
        Object.keys(INDEXS).forEach(function (key) {
            data = data.concat(Object.keys(INDEXS[key]).map(function (page) {
                return INDEXS[key][page];
            }));
        });

        query = query.trim();
        var keywords = query.split(/[\s\-，\\/]+/);
        if (keywords.length !== 1) {
            keywords = [].concat(query, keywords);
        }

        var loop = function (i) {
            var post = data[i];
            var matchesScore = 0;
            var resultStr = '';
            var handlePostTitle = '';
            var handlePostContent = '';
            var postTitle = post.title && post.title.trim();
            var postContent = post.body && post.body.trim();
            var postUrl = post.slug || '';

            if (postTitle) {
                keywords.forEach(function (keyword) {
                    // From https://github.com/sindresorhus/escape-string-regexp
                    var regEx = new RegExp(
                        escapeHtml(ignoreDiacriticalMarks(keyword)).replace(
                            /[|\\{}()[\]^$+*?.]/g,
                            '\\$&',
                        ),
                        'gi',
                    );
                    var indexTitle = -1;
                    var indexContent = -1;
                    handlePostTitle = postTitle
                        ? escapeHtml(ignoreDiacriticalMarks(postTitle))
                        : postTitle;
                    handlePostContent = postContent
                        ? escapeHtml(ignoreDiacriticalMarks(postContent))
                        : postContent;

                    indexTitle = postTitle ? handlePostTitle.search(regEx) : -1;
                    indexContent = postContent ? handlePostContent.search(regEx) : -1;

                    if (indexTitle >= 0 || indexContent >= 0) {
                        matchesScore += indexTitle >= 0 ? 3 : indexContent >= 0 ? 2 : 0;
                        if (indexContent < 0) {
                            indexContent = 0;
                        }

                        var start = 0;
                        var end = 0;

                        start = indexContent < 11 ? 0 : indexContent - 10;
                        end = start === 0 ? 70 : indexContent + keyword.length + 60;

                        if (postContent && end > postContent.length) {
                            end = postContent.length;
                        }

                        var matchContent =
                            '...' +
                            handlePostContent
                                .substring(start, end)
                                .replace(
                                    regEx,
                                    function (word) {
                                        return ('<em class="search-keyword">' + word + '</em>');
                                    },
                                ) +
                            '...';

                        resultStr += matchContent;
                    }
                });

                if (matchesScore > 0) {
                    var matchingPost = {
                        title: handlePostTitle,
                        content: postContent ? resultStr : '',
                        url: postUrl,
                        score: matchesScore,
                    };

                    matchingResults.push(matchingPost);
                }
            }
        };

        for (var i = 0; i < data.length; i++) loop(i);

        return matchingResults.sort(function (r1, r2) {
            return r2.score - r1.score;
        });
    }

    function init(config, vm) {
        var isAuto = config.paths === 'auto';
        var paths = isAuto ? getAllPaths(vm.router) : config.paths;

        var namespaceSuffix = '';

        // only in auto mode
        if (paths.length && isAuto && config.pathNamespaces) {
            var path = paths[0];

            if (Array.isArray(config.pathNamespaces)) {
                namespaceSuffix =
                    config.pathNamespaces.filter(
                        function (prefix) {
                            return path.slice(0, prefix.length) === prefix;
                        },
                    )[0] || namespaceSuffix;
            } else if (config.pathNamespaces instanceof RegExp) {
                var matches = path.match(config.pathNamespaces);

                if (matches) {
                    namespaceSuffix = matches[0];
                }
            }
            var isExistHome = paths.indexOf(namespaceSuffix + '/') === -1;
            var isExistReadme = paths.indexOf(namespaceSuffix + '/README') === -1;
            if (isExistHome && isExistReadme) {
                paths.unshift(namespaceSuffix + '/');
            }
        } else if (paths.indexOf('/') === -1 && paths.indexOf('/README') === -1) {
            paths.unshift('/');
        }

        var expireKey = resolveExpireKey(config.namespace) + namespaceSuffix;
        var indexKey = resolveIndexKey(config.namespace) + namespaceSuffix;

        var isExpired = localStorage.getItem(expireKey) < Date.now();

        INDEXS = JSON.parse(localStorage.getItem(indexKey));

        if (isExpired) {
            INDEXS = {};
        } else if (!isAuto) {
            return;
        }

        var len = paths.length;
        var count = 0;

        paths.forEach(function (path) {
            if (INDEXS[path]) {
                return count++;
            }

            Docsify.get(vm.router.getFile(path), false, vm.config.requestHeaders).then(
                function (result) {
                    INDEXS[path] = genIndex(path, result, vm.router, config.depth);
                    len === ++count && saveData(config.maxAge, expireKey, indexKey);
                },
            );
        });
    }

    /* eslint-disable no-unused-vars */

    var NO_DATA_TEXT = '';
    var options;

    function style() {
        var code = '\n.sidebar {\n  padding-top: 0;\n}\n\n.search {\n  margin-bottom: 20px;\n  padding: 6px;\n  border-bottom: 1px solid #eee;\n}\n\n.search .input-wrap {\n  display: flex;\n  align-items: center;\n}\n\n.search .results-panel {\n  display: none;\n}\n\n.search .results-panel.show {\n  display: block;\n}\n\n.search input {\n  outline: none;\n  border: none;\n  width: 100%;\n  padding: 0 7px;\n  line-height: 36px;\n  font-size: 14px;\n  border: 1px solid transparent;\n}\n\n.search input:focus {\n  box-shadow: 0 0 5px var(--theme-color, #42b983);\n  border: 1px solid var(--theme-color, #42b983);\n}\n\n.search input::-webkit-search-decoration,\n.search input::-webkit-search-cancel-button,\n.search input {\n  -webkit-appearance: none;\n  -moz-appearance: none;\n  appearance: none;\n}\n.search .clear-button {\n  cursor: pointer;\n  width: 36px;\n  text-align: right;\n  display: none;\n}\n\n.search .clear-button.show {\n  display: block;\n}\n\n.search .clear-button svg {\n  transform: scale(.5);\n}\n\n.search h2 {\n  font-size: 17px;\n  margin: 10px 0;\n}\n\n.search a {\n  text-decoration: none;\n  color: inherit;\n}\n\n.search .matching-post {\n  border-bottom: 1px solid #eee;\n}\n\n.search .matching-post:last-child {\n  border-bottom: 0;\n}\n\n.search p {\n  font-size: 14px;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  display: -webkit-box;\n  -webkit-line-clamp: 2;\n  -webkit-box-orient: vertical;\n}\n\n.search p.empty {\n  text-align: center;\n}\n\n.app-name.hide, .sidebar-nav.hide {\n  display: none;\n}';

        Docsify.dom.style(code);
    }

    function tpl(defaultValue) {
        if (defaultValue === void 0) defaultValue = '';

        var html = '<div class="input-wrap">\n      <input type="search" value="' + defaultValue + '" aria-label="Search text" />\n      <div class="clear-button">\n        <svg width="26" height="24">\n          <circle cx="12" cy="12" r="11" fill="#ccc" />\n          <path stroke="white" stroke-width="2" d="M8.25,8.25,15.75,15.75" />\n          <path stroke="white" stroke-width="2"d="M8.25,15.75,15.75,8.25" />\n        </svg>\n      </div>\n    </div>\n    <div class="results-panel"></div>\n    </div>';
        var el = Docsify.dom.create('div', html);
        var aside = Docsify.dom.find('aside');

        Docsify.dom.toggleClass(el, 'search');
        Docsify.dom.before(aside, el);
    }

    function doSearch(value) {
        var $search = Docsify.dom.find('div.search');
        var $panel = Docsify.dom.find($search, '.results-panel');
        var $clearBtn = Docsify.dom.find($search, '.clear-button');
        var $sidebarNav = Docsify.dom.find('.sidebar-nav');
        var $appName = Docsify.dom.find('.app-name');

        if (!value) {
            $panel.classList.remove('show');
            $clearBtn.classList.remove('show');
            $panel.innerHTML = '';

            if (options.hideOtherSidebarContent) {
                $sidebarNav && $sidebarNav.classList.remove('hide');
                $appName && $appName.classList.remove('hide');
            }

            return;
        }

        var matchs = search(value);

        var html = '';
        matchs.forEach(function (post) {
            html += '<div class="matching-post">\n<a href="' + (post.url) + '">\n<h2>' + (post.title) + '</h2>\n<p>' + (post.content) + '</p>\n</a>\n</div>';
        });

        $panel.classList.add('show');
        $clearBtn.classList.add('show');
        $panel.innerHTML = html || ('<p class="empty">' + NO_DATA_TEXT + '</p>');
        if (options.hideOtherSidebarContent) {
            $sidebarNav && $sidebarNav.classList.add('hide');
            $appName && $appName.classList.add('hide');
        }
    }

    function bindEvents() {
        var $search = Docsify.dom.find('div.search');
        var $input = Docsify.dom.find($search, 'input');
        var $inputWrap = Docsify.dom.find($search, '.input-wrap');

        var timeId;

        /**
         Prevent to Fold sidebar.

         When searching on the mobile end,
         the sidebar is collapsed when you click the INPUT box,
         making it impossible to search.
         */
        Docsify.dom.on(
            $search,
            'click',
            function (e) {
                return [ 'A', 'H2', 'P', 'EM' ].indexOf(e.target.tagName) === -1 &&
                    e.stopPropagation();
            },
        );
        Docsify.dom.on($input, 'input', function (e) {
            clearTimeout(timeId);
            timeId = setTimeout(function (_) {
                return doSearch(e.target.value.trim());
            }, 100);
        });
        Docsify.dom.on($inputWrap, 'click', function (e) {
            // Click input outside
            if (e.target.tagName !== 'INPUT') {
                $input.value = '';
                doSearch();
            }
        });
    }

    function updatePlaceholder(text, path) {
        var $input = Docsify.dom.getNode('.search input[type="search"]');

        if (!$input) {
            return;
        }

        if (typeof text === 'string') {
            $input.placeholder = text;
        } else {
            var match = Object.keys(text).filter(function (key) {
                return path.indexOf(key) > -1;
            })[0];
            $input.placeholder = text[match];
        }
    }

    function updateNoData(text, path) {
        if (typeof text === 'string') {
            NO_DATA_TEXT = text;
        } else {
            var match = Object.keys(text).filter(function (key) {
                return path.indexOf(key) > -1;
            })[0];
            NO_DATA_TEXT = text[match];
        }
    }

    function updateOptions(opts) {
        options = opts;
    }

    function init$1(opts, vm) {
        var keywords = vm.router.parse().query.s;

        updateOptions(opts);
        style();
        tpl(keywords);
        bindEvents();
        keywords && setTimeout(function (_) {
            return doSearch(keywords);
        }, 500);
    }

    function update(opts, vm) {
        updateOptions(opts);
        updatePlaceholder(opts.placeholder, vm.route.path);
        updateNoData(opts.noData, vm.route.path);
    }

    /* eslint-disable no-unused-vars */

    var CONFIG = {
        placeholder: 'Type to search',
        noData: 'No Results!',
        paths: 'auto',
        depth: 2,
        maxAge: 86400000, // 1 day
        hideOtherSidebarContent: false,
        namespace: undefined,
        pathNamespaces: undefined,
    };

    var install = function (hook, vm) {
        var util = Docsify.util;
        var opts = vm.config.search || CONFIG;

        if (Array.isArray(opts)) {
            CONFIG.paths = opts;
        } else if (typeof opts === 'object') {
            CONFIG.paths = Array.isArray(opts.paths) ? opts.paths : 'auto';
            CONFIG.maxAge = util.isPrimitive(opts.maxAge) ? opts.maxAge : CONFIG.maxAge;
            CONFIG.placeholder = opts.placeholder || CONFIG.placeholder;
            CONFIG.noData = opts.noData || CONFIG.noData;
            CONFIG.depth = opts.depth || CONFIG.depth;
            CONFIG.hideOtherSidebarContent =
                opts.hideOtherSidebarContent || CONFIG.hideOtherSidebarContent;
            CONFIG.namespace = opts.namespace || CONFIG.namespace;
            CONFIG.pathNamespaces = opts.pathNamespaces || CONFIG.pathNamespaces;
        }

        var isAuto = CONFIG.paths === 'auto';

        hook.mounted(function (_) {
            init$1(CONFIG, vm);
            !isAuto && init(CONFIG, vm);
        });
        hook.doneEach(function (_) {
            update(CONFIG, vm);
            isAuto && init(CONFIG, vm);
        });
    };

    $docsify.plugins = [].concat(install, $docsify.plugins);

}());